### ISSUE TITLE
Kubelet makes excessive calls to EC2 metadata service, which has rate limits

### ISSUE THREAD

**Issue by PaulFurtado** [kubernetes/kubernetes#110136]
### What happened?

We frequently see AWS metadata service rate limiting on kube nodes and one of the worst offenders is kubelet. if not aware, the AWS IMDS `169.254.169.254` has a packet-based rate limit of 1024 packets per second which includes both IMDS and DNS.

On the node I'm currently investigating, every 10 seconds, it makes 36 useless requests which amounts to a few hundred packets amounting to about a quarter of the rate limit for that second.

The specific requests are:
```
GET /latest/meta-data/network/interfaces/macs/ HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:05:07:5e:75:10/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:17:15:2b:68:46/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:34:d4:f3:96:22/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:4d:38:5b:ec:f6/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:89:07:a6:51:5c/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:90:48:a8:10:fe/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:a0:32:b3:0d:a2/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:cd:0d:e7:41:58/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:d1:72:8b:42:fe/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:de:87:a6:e6:cc/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:e2:92:67:48:b0/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:eb:f3:69:4c:4e/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:f3:98:f1:f1:9c/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:fb:58:15:2a:dc/device-number HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:d1:72:8b:42:fe/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:fb:58:15:2a:dc/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:4d:38:5b:ec:f6/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:e2:92:67:48:b0/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:eb:f3:69:4c:4e/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:a0:32:b3:0d:a2/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:de:87:a6:e6:cc/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:90:48:a8:10:fe/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:34:d4:f3:96:22/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:17:15:2b:68:46/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:f3:98:f1:f1:9c/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:89:07:a6:51:5c/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:05:07:5e:75:10/local-ipv4s HTTP/1.1
GET /latest/meta-data/network/interfaces/macs/0a:cd:0d:e7:41:58/local-ipv4s HTTP/1.1
GET /latest/meta-data/public-ipv4 HTTP/1.1
GET /latest/meta-data/public-ipv4 HTTP/1.1
GET /latest/meta-data/public-ipv4 HTTP/1.1
GET /latest/meta-data/local-hostname HTTP/1.1
GET /latest/meta-data/public-hostname HTTP/1.1
GET /latest/meta-data/public-hostname HTTP/1.1
GET /latest/meta-data/public-hostname HTTP/1.1
```

The device-number and local-ipv4s are particularly useless because kubelet only cares about the device 0 IP addresses and ignores the rest. A trivial optimization there is to iterate the mac addresses until device 0 is found, and then only make the `local-ipv4s` call for that mac.

That said, the private ipv4 address for eth0 cannot possibly change while an instance is running, the instance must be stopped and started. So for the life of a kubelet process, it only needs to fetch the ipv4 address a single time. The same is actually true for the private hostname too since it is IP based. It is now possible to make it instance-id based with a launch setting as well, but that is also not going to change for the life of the kubelet process.

So most of these requests could be made a single time on startup. The only ones that can possibly change without a reboot are the public values. If for some reason it HAD to poll frequently though, it could probably reduce this all to a single request for the local-hostname and then parse the IP out of it if it is in the `ip-172.16.1.100` format, but making these requests on startup a single time is less hacky.

For the public-ipv4 and the public-hostname case, you can see that it makes 3 requests each time. That's because on an instance without a public IP, the endpoint returns 404 and the client is retrying the 404 three times instead of treating 404 as unretryable. I also can't imagine someone actually changing an instnace's public IP while running, but I guess if they do it definitely makes sense for kubelet to handle it. That said, polling every 10 seconds for it still seems excessive since surely that would be a very rare event.

### What did you expect to happen?

Kubelet shouldn't spam the AWS metadata service every 10 seconds.

### How can we reproduce it (as minimally and precisely as possible)?

Run a kubelet on AWS and use tcpdump to monitor the requests it makes.

### Kubernetes version

1.21.12

### Cloud provider

AWS

### OS version

CentOS 8 Stream

### Install tools

in-house

### Container runtime (CRI) and version (if applicable)

NA

### Related plugins (CNI, CSI, ...) and versions (if applicable)

NA

**Comment by ialidzhikov** [comment]
/area provider/aws
cc @nckturner @hakman

**Comment by sftim** [comment]
> the endpoint returns 404 and the client is retrying the 404 three times

Even if we did want to retry, shouldn't that involve a back-off?

**Comment by nckturner** [comment]
Yes, it sounds like it could be optimized, but is it worth spending any effort on this?  The external cloud provider does not query the instance metadata service for node addresses, so if its not causing too much pain, the answer is probably to wait for upgrade.  Otherwise, fixing the 404 retry (or polling less frequently, depending where the code change is) could be easy stop-gap solutions.

**Comment by nckturner** [comment]
Given this will not be a problem after migration, I'm going to close it unless there is some imminent need.  Please feel to reopen if that's the case. 

/close
